### PR TITLE
[SPARK-28189][SQL] Use semanticEquals in Dataset drop method for attributes comparison

### PR DESCRIPTION
**Pull request by Tonix517**
## What changes were proposed in this pull request?

In Dataset drop(col: Column) method, the `equals` comparison method was used instead of `semanticEquals`, which caused the problem of abnormal case-sensitivity behavior. When attributes of LogicalPlan are checked for equality, `semanticEquals` should be used instead.

A similar PR I referred to: https://github.com/apache/spark/pull/22713 created by @mgaido91 

## How was this patch tested?

- Added new unit test case in DataFrameSuite
- ./build/sbt "testOnly org.apache.spark.sql.*"
- The python code from ticket reporter at https://issues.apache.org/jira/browse/SPARK-28189

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -572,6 +572,22 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     assert(df.schema.map(_.name) === Seq("value"))
   }
 
+  test("drop column using drop with column reference with case-insensitive names") {
+    val col1 = testData("KEY")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Could you check this on both cases: spark.sql.caseSensitive=true/false?

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 2]
```diff
@@ -2322,7 +2322,7 @@ class Dataset[T] private[sql](
     }
     val attrs = this.logicalPlan.output
     val colsAfterDrop = attrs.filter { attr =>
-      attr != expression
+      !attr.semanticEquals(expression)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Is there no palce having the same issue?

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 3]
```diff
@@ -572,6 +572,22 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+    val df1 = testData.drop(col1)
+    checkAnswer(
+      df1,
+      testData.collect().map(x => Row(x.getString(1))).toSeq)
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: `checkAnswer(df1, testData.selectExpr("value"))`?

**Comment by maropu** [comment]
ok to test

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: +1 for @maropu 's comment.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 4]
```diff
@@ -572,6 +572,22 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     assert(df.schema.map(_.name) === Seq("value"))
   }
 
+  test("drop column using drop with column reference with case-insensitive names") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Please use `SPARK-28189` prefix at the bug fix test case.
```
-  test("drop column using drop with column reference with case-insensitive names") {
+  test("SPARK-28189 drop column using drop with column reference with case-insensitive names") {
```

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (Tonix517) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: With that flag set as `true`, an exception was thrown out `org.apache.spark.sql.AnalysisException: Cannot resolve column name "KEY" among (key, value);`. It looks to be the correct behavior in that case. do I still need to check it in this test case and make sure such exception is thrown and caught?

**Comment by Tonix517** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (Tonix517) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Went through Dataset.scala - didn't find similar issue. However there might be the same problems in other places in our SQL code..

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: There are other places. Please see https://github.com/apache/spark/pull/21449.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Yep. For case-sensitive issue, we need to have a test coverage for both.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (tugangkai) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nice

**Comment by Tonix517** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for update, @Tonix517 .

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: ok, thanks. I think its ok to only target dataset.drop in this pr.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: +1 for @maropu 's opinion.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for your first contribution, @Tonix517 .
Thank you, @maropu and @mgaido91 , too.
Merged to `master`.

**Comment by Tonix517** [comment]
@dongjoon-hyun @maropu @mgaido91  thank you guys very much on helping my first commit - looking forward to working with you more :)

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 5]
```diff
@@ -572,6 +572,29 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
...
+      val col2 = testData("Key")
+      val df2 = testData.drop(col2)
+      checkAnswer(df2, testData.selectExpr("value"))
+      assert(df2.schema.map(_.name) === Seq("value"))
+    }
+
+    // With SQL config caseSensitive ON, AnalysisException should be thrown
+    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "true") {
+      val e = intercept[AnalysisException] {
+        testData("KEY")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: wait, it seems this test is not related to this pr?
you just wanted to do like this?
```
  test("SPARK-28189 drop column using drop with column reference with case-insensitive names") {
    var caseInsensitiveCol: Column = null
    withSQLConf(SQLConf.CASE_SENSITIVE.key -> "false") {
      caseInsensitiveCol = testData("KEY")
    }

    Seq("true", "false").foreach { isCaseSensitive =>
      withSQLConf(SQLConf.CASE_SENSITIVE.key -> isCaseSensitive) {
        caseInsensitiveCol = testData("KEY")
        val df = testData.drop(caseInsensitiveCol)
        checkAnswer(df, testData.selectExpr("value"))
        assert(df.schema.map(_.name) === Seq("value"))
      }
    }
  }
```

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: ? @maropu . In your example, the following will fail in the same way because `testData` only has `key` and `value`.
```
      withSQLConf(SQLConf.CASE_SENSITIVE.key -> isCaseSensitive) {
        caseInsensitiveCol = testData("KEY")
```
For me, the test case looked okay because `caseSensitive` mode doesn't allow that kind of unmatched column from the beginning.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: oh, sorry and my bad. That line wasn't needed and I updated the example above.
Anyway, I just a bit worry about the behaivour change in the code flow below;
```
scala> val testData = Seq(("a", 1)).toDF("key", "value")

scala> sql("SET spark.sql.caseSensitive=false")
scala> val caseInsensitiveCol = testData("KEY")

scala> sql("SET spark.sql.caseSensitive=true")
scala> testData.drop(caseInsensitiveCol).show()
// v2.4.3
+---+-----+
|key|value|
+---+-----+
|  a|    1|
+---+-----+

// master w/this pr
+-----+
|value|
+-----+
|    1|
+-----+
```

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: First, it's not a correct(?) use-case to switch the case sensitivity between declaring the variable and using the variable. 
Second, I merged this to `master` only with the similar concern. :)
Since this is a bug fix, I think we don't need a documentation about this. And, 3.0.0 is a good place to change those behavior change bug fix.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: ok, thanks for the check!

Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Do you have any suggestion for your concern? Then, please share it with us~ You're welcome.

Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: NVM. I just a bit worry about not use cases but the test coverage.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Yes. This test case of the second part (SQLConf.CASE_SENSITIVE.key -> "true") is weird to me.

Comment (gatorsmile) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: Anyway, we can keep it unchanged. Ideally, we do not need the second part.

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]:
```
  def semanticEquals(other: Expression): Boolean =
    deterministic && other.deterministic && canonicalized == other.canonicalized
```

What is the reason the comparison should be related to the deterministic when we want to drop it?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: nvm. The output only contains Attribute

**Review by gatorsmile — COMMENTED** [review]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 5 above)
Comment (maropu) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: ok, I made a pr to drop that part: https://github.com/apache/spark/pull/25216
anyway, thanks for the check!
